src/ps.c

```diff
@@ -0,0 +1,336 @@
+#include "bouncer.h"
+
+#include <usual/crypto/csrandom.h>
+#include <usual/hashing/spooky.h>
+#include <usual/hashtab-impl.h>
+
+static PgParsedPreparedStatement *create_prepared_statement(PgParsePacket *parse_packet)
+{
+  static bool initialized;
+	static uint32_t rand_seed;
+  PgParsedPreparedStatement *s = NULL;
+
+	if (!initialized) {
+		initialized = true;
+		rand_seed = csrandom();
+	}
+
+  s = (PgParsedPreparedStatement *)malloc(sizeof(*s));
+  s->name = parse_packet->name;
+  s->pkt = parse_packet;
+  s->query_hash[0] = rand_seed;
+	s->query_hash[1] = 0;
+	spookyhash(parse_packet->query, strlen(parse_packet->query), &s->query_hash[0], &s->query_hash[1]);
```

Comment:
I wonder why don't you take in account hash collisions here?
Yes, probability of collision of spooky hash is assumed to be quite small, but still it is non-zero.
And if it happens, then pgbouncer locates wrong query and application receives wrong result.